Build failure on macOS 10.15: `error: no viable constructor or deduction guide for deduction of template arguments of 'function'`

### Description

```
[ 95%] Building CXX object test/CMakeFiles/ssl_tests.dir/ssl_tests.cpp.o
cd /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/build/test && /usr/bin/clang++ -DCPR_BUILD_TESTS -DCPR_BUILD_TESTS_SSL -DCPR_ENABLE_CURL_HTTP_ONLY -DCPR_ENABLE_SSL -DCPR_FORCE_OPENSSL_BACKEND -DCPR_USE_SYSTEM_CURL -DCPR_USE_SYSTEM_GTEST -DMG_ENABLE_OPENSSL -DOPENSSL_BACKEND_USED -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/build/_deps/mongoose-src -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/include -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/build/cpr_generated_includes -isystem /opt/local/include -pipe -I/opt/local/libexec/openssl3/include -Os -Wno-error=overflow -DNDEBUG -I/opt/local/libexec/openssl3/include -isystem/opt/local/include -stdlib=libc++ -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -Wall -Wextra -Wpedantic -Werror -std=gnu++17 -arch x86_64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.15 -MD -MT test/CMakeFiles/ssl_tests.dir/ssl_tests.cpp.o -MF CMakeFiles/ssl_tests.dir/ssl_tests.cpp.o.d -o CMakeFiles/ssl_tests.dir/ssl_tests.cpp.o -c /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/ssl_tests.cpp
/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:316:25: error: no viable constructor or deduction guide for deduction of template arguments of 'function'
    const std::function observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
                        ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate template ignored: could not match 'function<_Fp>' against '(lambda at /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:316:37)'
template<class _Fp> class _LIBCPP_TEMPLATE_VIS function; // undefined
                                               ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate function template not viable: requires 0 arguments, but 1 was provided
/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:345:25: error: no viable constructor or deduction guide for deduction of template arguments of 'function'
    const std::function observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
                        ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate template ignored: could not match 'function<_Fp>' against '(lambda at /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:345:37)'
template<class _Fp> class _LIBCPP_TEMPLATE_VIS function; // undefined
                                               ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate function template not viable: requires 0 arguments, but 1 was provided
/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:377:25: error: no viable constructor or deduction guide for deduction of template arguments of 'function'
    const std::function observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
                        ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate template ignored: could not match 'function<_Fp>' against '(lambda at /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:377:37)'
template<class _Fp> class _LIBCPP_TEMPLATE_VIS function; // undefined
                                               ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate function template not viable: requires 0 arguments, but 1 was provided
/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:418:25: error: no viable constructor or deduction guide for deduction of template arguments of 'function'
    const std::function observer_fn{[N](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
                        ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate template ignored: could not match 'function<_Fp>' against '(lambda at /opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_devel_cpr/cpr/work/cpr-1.10.5/test/multiasync_tests.cpp:418:37)'
template<class _Fp> class _LIBCPP_TEMPLATE_VIS function; // undefined
                                               ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/functional:1430:48: note: candidate function template not viable: requires 0 arguments, but 1 was provided
```

Buildbot log: https://build.macports.org/builders/ports-10.15_x86_64-builder/builds/192100/steps/install-port/logs/stdio

### Example/How to Reproduce

On Catalina run:
```
sudo port -v build cpr
```

### Possible Fix

_No response_

### Where did you get it from?

Other (specify in "Additional Context/Your Environment")

### Additional Context/Your Environment

- OS: macOS
- Version: 10.15
- MacPorts

--- a/test/multiasync_tests.cpp
+++ b/test/multiasync_tests.cpp
@@ -1,3 +1,4 @@
+#include <functional>
 #include <string>
 #include <vector>
 
@@ -313,7 +314,7 @@ static TestSynchronizationEnv* synchro_env = new TestSynchronizationEnv();
 TEST(MultiAsyncCancelTests, CancellationOnQueue) {
     synchro_env->Reset();
     const Url hello_url{server->GetBaseUrl() + "/hello.html"};
-    const std::function observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
+    const std::function<bool(cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t)> observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
         synchro_env->fn_called.store(true);
         return true;
     }};
@@ -342,7 +343,7 @@ TEST(MultiAsyncCancelTests, TestCancellationInTransit) {
 
     // 1. Thread running the test acquires the condition variable's mutex
     std::unique_lock setup_lock{synchro_env->test_cv_mutex};
-    const std::function observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
+    const std::function<bool(cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t)> observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
         if (synchro_env->counter == 0) {
             // 3. in Threadpool, the cv mutex is obtained by the worker thread
             const std::unique_lock l{synchro_env->test_cv_mutex};
@@ -374,7 +375,7 @@ TEST(MultiAsyncCancelTests, TestCancellationOnResponseWrapperDestruction) {
     const Url call_url{server->GetBaseUrl() + "/hello.html"};
     synchro_env->Reset();
     std::unique_lock setup_lock{synchro_env->test_cv_mutex};
-    const std::function observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
+    const std::function<bool(cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t)> observer_fn{[](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
         const std::unique_lock l{synchro_env->test_cv_mutex};
         synchro_env->counter++;
         synchro_env->test_cv.notify_all();
@@ -415,7 +416,7 @@ TEST(MultiAsyncCancelTests, TestIntervalOfProgressCallsLowSpeed) {
     // This variable will be used to cancel the transaction at the point of the Nth call.
     const std::chrono::time_point start{std::chrono::steady_clock::now()};
 
-    const std::function observer_fn{[N](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
+    const std::function<bool(cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t)> observer_fn{[N](cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, cpr_pf_arg_t, intptr_t) -> bool {
         const size_t current_iteration{++(synchro_env->counter)};
         return current_iteration <= N;
     }};

